### PR TITLE
roll: conservatively treat unknown-health nodes as healthy; support capacity increase

### PR DESCRIPTION
**Pull request by petertseng**
We wish to be conservative because we don't want to mistakenly take down
a node of unknown health only to find out that it is healthy. The other
advantage of this is that it supports easily migrating apps from other
deployment systems which may not report health.

### CODE REVIEW

**Comment by anthonybishopric** [comment]
Looks good! May want to write a test for `shouldUpdateAfterRoll`, and possibly the main loop if possible.

**Comment by anthonybishopric** [comment]
*`shouldRollAfterDelay`

File: pkg/roll/run_update.go [hunk 1]
```diff
@@ -360,10 +363,18 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
...
+			if hres.Status == health.Passing {
+				ret.Healthy++
+			} else if hres.Status != health.Unknown {
+				ret.Unhealthy++
+			}
 		}
 	}
+
+	// We set unknown here to deal with the case of nodes not being in CurrentPods()
+	ret.Unknown = ret.Desired - ret.Healthy - ret.Unhealthy
```
Comment (petertseng) [inline pkg/roll/run_update.go]: check whether this should be `ret.Real - ret.Healthy - ret.Unhealthy` instead. See discussion in comment on test

File: pkg/roll/update_test.go [hunk 2]
```diff
@@ -14,6 +21,138 @@ import (
...
+		"node3": {Status: health.Critical},
+	}
+	counts, err := upd.countHealthy(id, checks)
+	Assert(t).IsNil(err, "expected no error counting health")
+	expected := rcNodeCounts{
+		Desired: 3,
+		Current: 3,
+		Real:    2,
+		Healthy: 2,
+		Unknown: 1,
```
Comment (petertseng) [inline pkg/roll/update_test.go]: in this state, should `Unknown` be 0 or 1? Need to think about this for a bit.

File: pkg/roll/run_update.go (hunk 1 above)
Comment (petertseng) [inline pkg/roll/run_update.go]: upon further deliberation, I think that would be incorrect. I believe this line should stand as written. A new test case will be added that will need this line to subtract from `Desired`, not `Real`

**Comment by petertseng** [comment]
Need to add a test for the case where number of nodes belonging to the RC is less than the desired number (like the case when converting from a different deployment system)

**Comment by petertseng** [comment]
Along with the extra test to `countHealthy`, tests for `shouldRollAfterUpdate` have been added. It does not appear to be feasible to test the `ROLL_LOOP` at this time (unless an extraction is performed) since it requires significant setup and various channels., but since fundamentally it is using the same `rollAlgorithm(oldNodes.Healthy+oldNodes.Unknown, newNodes.Healthy, u.DesiredReplicas, u.MinimumReplicas)` line as in `shouldRollAfterUpdate`, that means that the behavior I want to test has already been covered and any other tests would be outside the scope of this PR

**Comment by petertseng** [comment]
For your consideration are new commits added to this PR that supports the case of increasing capacity in conjunction with a rolling update.

The use case we have is that when we are converting from a different deployment system, we create two RCs, both with desired replicas 0. Then we increase capacity with a rolling update. Details in the individual commit messages.

File: pkg/roll/update_test.go [hunk 3]
```diff
@@ -6,7 +6,15 @@ import (
 	"testing"
```
Comment (mpuncel) [inline pkg/roll/update_test.go]: :bow: thanks for being rigorous about this

File: pkg/roll/update_test.go [hunk 4]
```diff
@@ -14,15 +22,27 @@ import (
...
+	Assert(t).AreEqual(uniformRollAlgorithm(t, 2, 2, 6, 3), 1, "should schedule difference if minimum must be maintained")
+	Assert(t).AreEqual(uniformRollAlgorithm(t, 1, 3, 6, 3), 3, "should schedule remaining if minimum is satisfied by new")
+	Assert(t).AreEqual(uniformRollAlgorithm(t, 3, 0, 6, 0), 6, "should schedule remaining if no minimum")
+}
 
-	Assert(t).AreEqual(rollAlgorithm(2, 2, 6, 3), 1, "should schedule difference if minimum must be maintained")
-	Assert(t).AreEqual(rollAlgorithm(1, 3, 6, 3), 3, "should schedule remaining if minimum is satisfied by new")
-	Assert(t).AreEqual(rollAlgorithm(3, 0, 6, 0), 6, "should schedule remaining if no minimum")
+func TestRollAlgorithmIncreases(t *testing.T) {
+	//Assert(t).AreEqual(rollAlgorithm(0, 0, 0, 3, 2), 1, "should schedule difference if increasing capacity from zero")
```
Comment (mpuncel) [inline pkg/roll/update_test.go]: are you intentionally leaving these around or should they be uncommented?

Comment (petertseng) [inline pkg/roll/update_test.go]: oops. those should be uncommented. just a moment.

File: pkg/roll/update_test.go [hunk 5]
```diff
@@ -138,19 +158,389 @@ func SimulateRollingUpgradeDisable(t *testing.T, full, nonew bool) {
...
 		// choose nodes from the eligible list, randomly, and put the new pod
 		// on them
-		for _, index := range rand.Perm(len(eligible))[:nextUpdate] {
+		for _, index := range rand.Perm(len(eligible))[:nextAdd] {
 			nodes[eligible[index]] = 1
 		}
 	}
 }
+
+type fakeHealthChecker struct {
```
Comment (mpuncel) [inline pkg/roll/update_test.go]: can you put this in a package like `pkg/health/checker/checkertest` or something like that? looks like it's generic enough to be useful (and people can implement what they need as they need it)

File: pkg/roll/update_test.go [hunk 6]
```diff
@@ -14,15 +22,27 @@ import (
...
 
+func uniformRollAlgorithm(t *testing.T, old, new, want, need int) int {
+	remove, add := rollAlgorithm(old, new, want, want, need)
+	Assert(t).AreEqual(remove, add, "expected nodes removed and nodes added to be equal")
+	return add
+}
+
 func TestWouldBlock(t *testing.T) {
 	// in the following cases, rollAlgorithm should return 0
-	Assert(t).AreEqual(rollAlgorithm(1, 1, 3, 4), 0, "should do nothing if below minimum")
```
Comment (petertseng) [inline pkg/roll/update_test.go]: this test was accidentally removed and was not readded. should fix that.

File: pkg/roll/update_test.go (hunk 5 above)
Comment (petertseng) [inline pkg/roll/update_test.go]: to my slight shame, there appears to already be one. I may simply add mine to this

Comment (petertseng) [inline pkg/roll/update_test.go]: oh wait, no, that was `checker_test.go` instead of `checkertest` package.

File: pkg/roll/update_test.go (hunk 4 above)
Comment (petertseng) [inline pkg/roll/update_test.go]: changed to real tests

File: pkg/roll/update_test.go (hunk 6 above)
Comment (petertseng) [inline pkg/roll/update_test.go]: readded.

File: pkg/roll/update_test.go (hunk 5 above)
Comment (petertseng) [inline pkg/roll/update_test.go]: added in 3c5192ce10c0ad0e93fed3b2a7515c48c0aa3a53 though I left a TODO that replication tests should probably put some stuff in there. maybe I'll just do that now instead of leaving the TODO

**Comment by petertseng** [comment]
splitting c376de4094ca0775f82b0ddd060217ad823a30dd into a commit for tests added merely to test existing behavior to make them separate from the tests that test new behavior. no code changes planned

File: pkg/roll/run_update.go (hunk 1 above)
Comment (anthonybishopric) [inline pkg/roll/run_update.go]: Why set `ret.Unknown` to this value instead of the sum of all healthchecks that `== health.Unknown`? Are we guarding against unknown unknowns?

File: pkg/roll/run_update.go [hunk 7]
```diff
@@ -360,84 +372,119 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
...
 	if err != nil {
-		return 0, util.Errorf("Could not determine old service health: %v", err)
+		return 0, 0, util.Errorf("Could not determine old service health: %v", err)
 	}
 
-	afterDelayNext := rollAlgorithm(afterDelayOld.Healthy, afterDelayNew.Healthy, u.DesiredReplicas, u.MinimumReplicas)
+	// We conservatively treat Unknown nodes as healthy on the old side.
+	afterDelayRemove, afterDelayNext := rollAlgorithm(
+		afterDelayOld.Healthy+afterDelayOld.Unknown, afterDelayNew.Healthy,
+		afterDelayOld.Desired+afterDelayNew.Desired, u.DesiredReplicas,
```
Comment (anthonybishopric) [inline pkg/roll/run_update.go]: Should these summary objects be parameters to `rollAlgorithm`? I ask because the decision to sum the health and unknown is copied to multiple locations in this file, and maybe we want tests to verify that it does the right thing.

File: pkg/roll/run_update.go (hunk 1 above)
Comment (petertseng) [inline pkg/roll/run_update.go]: We are. Should I take your question as an indication that the comment on the line above needs to be made clearer?

Comment (anthonybishopric) [inline pkg/roll/run_update.go]: More that I'd like to know why we have unknown unknowns. We should be explicit if we're guarding against something. Improving the comment to state exactly what cases we are trying to defend against would be a good first step.

File: pkg/roll/run_update.go [hunk 8]
```diff
@@ -360,84 +372,119 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
...
+//
+// Under the following circumstances, both return values will be zero:
 // - new >= desired (the update is done)
 // - old+new <= minHealthy (at or below the minimum, update has to block)
-func rollAlgorithm(old, new, desired, minHealthy int) int {
+func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (int, int) {
 	// how much "headroom" do we have between the number of nodes that are
 	// currently healthy, and the number that must be healthy?
 	// if we schedule more than this, we'll go below the minimum
 	headroom := old + new - minHealthy
```
Comment (petertseng) [inline pkg/roll/run_update.go]: add comment that this can go negative! and then `headroom += capacityIncrease` may increase that to just above positive, not the full value of `capacityIncrease`

Comment (petertseng) [inline pkg/roll/run_update.go]: A comment has been added in https://github.com/square/p2/pull/408/commits/01913a9c18a1733e765f336c565ad659e4fee426

Comment (petertseng) [inline pkg/roll/run_update.go]: added in aeb5b1fc8ab45241a64980efeaf77bfe5b9673fe

File: pkg/roll/run_update.go (hunk 7 above)
Comment (petertseng) [inline pkg/roll/run_update.go]: b627695084323a14eaccd60ae403b4fddf225ba7 consolidates the logic in one place

File: pkg/roll/run_update.go (hunk 1 above)
Comment (petertseng) [inline pkg/roll/run_update.go]: 0d88762ba6967c3e87c8e211404b28541a598233 now makes explicit the various cases where `Unknown` would get incremented. Take a look.

**Comment by petertseng** [comment]
I believe that addresses the current comments (responses inline on the respective comments)

File: pkg/roll/run_update.go [hunk 9]
```diff
@@ -332,6 +334,13 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
 	}
 	ret.Current = len(currentPods)
 
+	if ret.Desired > ret.Current {
+		// This implies that the RC hasn't yet scheuled pods that it desires to have.
```
Comment (petertseng) [inline pkg/roll/run_update.go]: s/scheuled/scheduled

File: pkg/health/checker/test/fake_checker.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package test
```
Comment (mpuncel) [inline pkg/health/checker/test/fake_checker.go]: thanks for doing this

File: pkg/roll/run_update.go [hunk 11]
```diff
@@ -398,6 +398,10 @@ func (u *update) shouldRollAfterDelay(newFields rcf.RC) (int, error) {
 	return afterDelayNext, nil
 }
 
+func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (int, int, int, int) {
```
Comment (mpuncel) [inline pkg/roll/run_update.go]: crafty

File: pkg/roll/run_update.go [hunk 12]
```diff
@@ -399,7 +399,8 @@ func (u *update) shouldRollAfterDelay(newFields rcf.RC) (int, error) {
 }
 
 func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (int, int, int, int) {
```
Comment (mpuncel) [inline pkg/roll/run_update.go]: nit: name the return parameters so it's more obvious what is happening now that there is some logic here

File: pkg/roll/run_update.go [hunk 13]
```diff
@@ -423,17 +426,35 @@ func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (int, in
 // Under the following circumstances, both return values will be zero:
 // - new >= desired (the update is done)
 // - old+new <= minHealthy (at or below the minimum, update has to block)
-func rollAlgorithm(old, new, desired, minHealthy int) (int, int) {
+func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (int, int) {
```
Comment (mpuncel) [inline pkg/roll/run_update.go]: nit: name the return params here too

File: pkg/roll/run_update.go [hunk 14]
```diff
@@ -455,7 +476,14 @@ func rollAlgorithm(old, new, desired, minHealthy int) (int, int) {
 	// remaining>0 and headroom>0. each one imposes an upper bound on how
 	// many we can schedule at once, so return the minimum of the two
```
Comment (mpuncel) [inline pkg/roll/run_update.go]: this comment seems irrelevant now

**Comment by mpuncel** [comment]
looks great, just some nits.

Thanks for writing comprehensive tests!!

File: pkg/roll/run_update.go (hunk 14 above)
Comment (petertseng) [inline pkg/roll/run_update.go]: Hmm, I think it is still relevant, though it's only accurate for nodes _added_. I'll see what I can do to bring it up to date.

Comment (mpuncel) [inline pkg/roll/run_update.go]: err yeah irrelevant was a strong word but it seems to suggest there is only one return value

**Comment by petertseng** [comment]
Named return variables and clarified the comment.

```diff
diff --git a/pkg/roll/run_update.go b/pkg/roll/run_update.go
index 7071aad..c6ef237 100644
--- a/pkg/roll/run_update.go
+++ b/pkg/roll/run_update.go
@@ -403,11 +403,13 @@ func (u *update) shouldRollAfterDelay(newFields rcf.RC) (int, int, error) {
        return afterDelayRemove, afterDelayAdd, nil
 }

-func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (int, int, int, int, int) {
-       // We conservatively treat Unknown nodes as healthy on the old side.
-       return oldHealth.Healthy + oldHealth.Unknown, newHealth.Healthy,
-               oldHealth.Desired + newHealth.Desired, u.DesiredReplicas,
-               u.MinimumReplicas
+func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (oldHealthy, newHealthy, currentDesired, targetDesired, minHealthy int) {
+       oldHealthy = oldHealth.Healthy + oldHealth.Unknown
+       newHealthy = newHealth.Healthy
+       currentDesired = oldHealth.Desired + newHealth.Desired
+       targetDesired = u.DesiredReplicas
+       minHealthy = u.MinimumReplicas
+       return
 }

 // the roll algorithm defines how to mutate RCs over time. it takes five args:
@@ -426,7 +428,7 @@ func (u *update) rollAlgorithmParams(oldHealth, newHealth rcNodeCounts) (int, in
 // Under the following circumstances, both return values will be zero:
 // - new >= desired (the update is done)
 // - old+new <= minHealthy (at or below the minimum, update has to block)
-func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (int, int) {
+func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (nodesToRemove, nodesToAdd int) {
        // how much "headroom" do we have between the number of nodes that are
        // currently healthy, and the number that must be healthy?
        // if we schedule more than this, we'll go below the minimum
@@ -474,11 +476,16 @@ func rollAlgorithm(old, new, currentDesired, targetDesired, minHealthy int) (int
                return 0, 0
        }
        // remaining>0 and headroom>0. each one imposes an upper bound on how
-       // many we can schedule at once, so return the minimum of the two
+       // many nodes we can add at once, so nodesToAdd is their minimum.
+       // nodesToRemove is the same number minus any capacity increase
+       // (so the net number of nodes added equals the capacity increase)
        if remaining < headroom {
-               return clampToZero(remaining - capacityIncrease), remaining
+               nodesToAdd = remaining
+       } else {
+               nodesToAdd = headroom
        }
-       return clampToZero(headroom - capacityIncrease), headroom
+       nodesToRemove = clampToZero(nodesToAdd - capacityIncrease)
+       return
 }

 func clampToZero(n int) int {
```

Where I deemed it appropriate to do so, I assigned variables to the named return values rather than using `return a, b, c`. The tests would have had my back if I had gotten it wrong

**Comment by petertseng** [comment]
I lost the comment. Putting it back.

**Comment by petertseng** [comment]
Rebased on master so as to make an upcoming test actually indicative of the state of the world

**Comment by petertseng** [comment]
tested in staging, correctly brings up a new service from capacity 0 -> 2

**Comment by mpuncel** [comment]
ship it!

**Comment by petertseng** [comment]
scheduled for ship after all problems revealed by https://github.com/square/p2/pull/414 have been fixed simply because I would much rather only try to reason about one change at once.

I may pull off some changes from this PR and use them in an upcoming bugfix.

**Comment by petertseng** [comment]
Rebase on master was necessary because the first four commits of this have been stolen by #426. I'll test in staging again for this reason, but I consider the previous approval to still be sufficient simply because the remaining three commits on this PR have not changed (except as necessary to deal with #426 's new behavior)

**Comment by petertseng** [comment]
If there is one place where more attention should be given, it's now that #426 interacts with 0ffe7f9987a432451a241fb7a394762c22848a3a in the following manner:

old healthy is now the minimum of (old desired, old healthy + old unknown).

I believe this is reasonable behavior.

**Comment by mpuncel** [comment]
still looks good :+1:

**Comment by anthonybishopric** [comment]
:shipit:
